stan/math/prim/meta/return_type.hpp

```diff
@@ -2,46 +2,205 @@
+ * @tparam Ts sequence of template parameter types
+ * @see return_type for a definition of the type ordering
+ * @ingroup type_trait
+ */
+template <typename... Ts>
+struct real_return {
+  using type = double;
+};
+
+template <typename T, typename... Ts>
+struct real_return<T, Ts...> {
+  using type
+      = promote_args_t<base_type_decay_t<T>, typename real_return<Ts...>::type>;
+};
+
+/**
+ * Convenience type to calculate the real return type.
  *
- * <code>return_type<...></code> is a class defining a single public
- * typedef <code>type</code> that is <code>var</code> if there is a non-constant
- * template argument and is <code>double</code> otherwise. This is consistent
- * with the Stan logic that if code receives <code>var</code> as an input type,
- * then the return type is <code>var</code>. All other functions return
- * <code>double</code>.
+ * @tparam Ts sequence of argument types
+ * @see real_return
+ * @ingroup type_trait
+ */
+template <typename... Ts>
+using real_return_t = typename real_return<Ts...>::type;
+
+/**
+ * Convenience type to calculate the complex return type, which wraps
+ * `std::complex` around the return type of the specified template parameters.
  *
- * Example usage:
- *  - <code>return_type<int,double,float>::type</code> is <code>double</code>
- *  - <code>return_type<double,var>::type</code> is <code>var</code>
+ * @tparam Ts sequence of argument types
+ * @see real_return
+ * @ingroup type_trait
+ */
+template <typename... Ts>
+using complex_return_t = std::complex<real_return_t<Ts...>>;
+
+/**
+ * Convenience type to calculate the complex return type, which wraps
+ * `std::vector` around the return type of the specified template parameters.
+ *
+ * @tparam Ts sequence of argument types
+ * @see real_return
+ * @ingroup type_trait
+ */
+template <typename... Ts>
+using std_vector_return_t = std::vector<real_return_t<Ts...>>;
+
+/**
+ * Convenience type to calculate the complex return type, which wraps
+ * `Eigen::Matrix< , -1, -1>` around the return type of the specified
+ * template parameters.
+ *
+ * @tparam Ts sequence of argument types
+ * @see real_return
+ * @ingroup type_trait
+ */
+template <typename... Ts>
+using matrix_return_t = Eigen::Matrix<real_return_t<Ts...>, -1, -1>;
+
+/**
+ * Convenience type to calculate the complex return type, which wraps
+ * `Eigen::Matrix< , -1, 1>` around the return type of the specified
+ * template parameters.
+ *
+ * @tparam Ts sequence of argument types
+ * @see real_return
+ * @ingroup type_trait
+ */
+template <typename... Ts>
+using vector_return_t = Eigen::Matrix<real_return_t<Ts...>, -1, 1>;
+
+/**
+ * Convenience type to calculate the complex return type, which wraps
+ * `Eigen::Matrix< , 1, -1>` around the return type of the specified
+ * template parameters.
+ *
+ * @tparam Ts sequence of argument types
+ * @see real_return
+ * @ingroup type_trait
+ */
+template <typename... Ts>
+using row_vector_return_t = Eigen::Matrix<real_return_t<Ts...>, 1, -1>;
```

Comment:
We can save this for later in case we find a use for something like this.  In general, I don't think we can predict anyting other than scalar or base return type given input types.